File: scripts/bundle-component-library.ts
import { join } from '@std/path/join';

export type BundleParams = {
    silent?: boolean;
};

function myLocation() {
    return new URL('.', import.meta.url).pathname;
}

const esbuildLocation = join(myLocation(), 'vendor', 'bin', 'esbuild');

async function bundleJavascriptWebComponentsLibrary(opts?: BundleParams) {
    const command = new Deno.Command(esbuildLocation, {
        args: ['--bundle', 'index.js', '--outfile=../static/index.bundle.js'],
        cwd: './src/website/components',
    });

    const result = await command.output();
    if (!result.success) {
        console.error(
            'Failed to bundle component library!',
            new TextDecoder().decode(result.stdout),
            new TextDecoder().decode(result.stderr),
        );
        throw new Error('Could not bundle component library');
    } else if (!opts?.silent) {
        console.log('Bundled component library into index.bundle.js');
    }
}

async function bundleCssStyles(opts?: BundleParams) {
    const command = new Deno.Command(esbuildLocation, {
        args: ['--bundle', 'index.css', '--outfile=index.bundle.css'],
        cwd: './src/website/static/css',
    });

    const result = await command.output();
    if (!result.success) {
        console.error(
            'Failed to bundle CSS styles!',
            new TextDecoder().decode(result.stdout),
            new TextDecoder().decode(result.stderr),
        );
        throw new Error('Could not bundle CSS styles!');
    } else if (!opts?.silent) {
        console.log('Bundled CSS files into index.bundle.css');
    }
}

export async function bundle(opts?: BundleParams) {
    await bundleJavascriptWebComponentsLibrary(opts);
    await bundleCssStyles(opts);
}

if (import.meta.main) {
    await bundle();
}

File: src/scripts/bundle_component_library.go
package main

import (
	"log"
	"os/exec"
	"path"
)

func runWithEsbuild(dir string, args []string) {
	cwd := GetCwdOrFail()
	esbuildPath := path.Join(cwd, "src", "scripts", "vendor", "bin", "esbuild")
	cmd := exec.Command(esbuildPath, args...)
	cmd.Dir = dir

	out, err := cmd.CombinedOutput()
	if err != nil {
		log.Panicf("Bundle failed %s\nError: %v", string(out), err)
	}
}

func bundleJavascriptWebcomponents(silent bool) {
	EnsureDirFromCwd("src", "goapp", "website", "static", "js")
	args := []string{
		"--bundle",
		"index.js",
		"--outfile=../static/js/index.bundle.js",
	}
	dir := path.Join(GetCwdOrFail(), "src", "goapp", "website", "components")
	runWithEsbuild(dir, args)
	if !silent {
		log.Println("Bundled component library into index.bundle.js")
	}
}

func bundleCss(silent bool) {
	args := []string{
		"--bundle",
		"index.css",
		"--loader:.svg=dataurl",
		"--outfile=index.bundle.css",
	}
	dir := path.Join(GetCwdOrFail(), "src", "goapp", "website", "static", "css")
	runWithEsbuild(dir, args)
	if !silent {
		log.Println("Bundling CSS files into index.bundle.css")
	}
}

func BundleComponentLibrarySilently() {
	bundleJavascriptWebcomponents(true)
	bundleCss(true)
}

func BundleComponentLibrary() {
	bundleJavascriptWebcomponents(false)
	bundleCss(false)
}
